### PR TITLE
Only GitLab runner can deploy to `sandbox` (#2842)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and build passes in `sandbox` <sub>or added `no sandbox` label</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Purchased BQ slot committment in `dev`                <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Deleted BQ slot committment in `dev`                  <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Purchased BQ slot committment in `prod`               <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Deleted BQ slot committment in `prod`                 <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by noah-aviel-dove** [comment]
Skipping integration tests as there's no way this can possibly affect them.

**Review by danielsotirhos — APPROVED** [review]
Approved.

```
(azul2) daniel@Crispin ~/azul2 $ _select
lrwxr-xr-x  1 daniel  staff  12 Apr  1 12:58 .active -> daniel.local
(azul2) daniel@Crispin ~/azul2 $
(azul2) daniel@Crispin ~/azul2 $ python scripts/check_branch.py
(azul2) daniel@Crispin ~/azul2 $

(azul2) daniel@Crispin ~/azul2 $ _select
lrwxr-xr-x  1 daniel  staff  7 Apr  1 12:58 .active -> sandbox
(azul2) daniel@Crispin ~/azul2 $
(azul2) daniel@Crispin ~/azul2 $ python scripts/check_branch.py
Traceback (most recent call last):
  File "scripts/check_branch.py", line 122, in <module>
    main(sys.argv[1:])
  File "scripts/check_branch.py", line 115, in main
    check_gitlab(stage, is_gitlab=is_gitlab)
  File "scripts/check_branch.py", line 72, in check_gitlab
    raise RuntimeError(f"Only the GitLab runner should deploy to '{stage}'")
RuntimeError: Only the GitLab runner should deploy to 'sandbox'
(azul2) daniel@Crispin ~/azul2 $
```

File: scripts/check_branch.py [hunk 1]
```diff
@@ -54,20 +55,37 @@ def check_branch(branch: Optional[str], stage: str) -> None:
...
+    if stage == 'sandbox' and not is_gitlab:
+        raise RuntimeError(f"Only the GitLab runner should deploy to '{stage}'")
+
+
 def expected_stage(branch: Optional[str]) -> Optional[str]:
     return config.main_deployments_by_branch.get(branch)
 
 
-def current_branch() -> Optional[str]:
+def current_branch() -> Tuple[Optional[str], bool]:
```
[the remark this answers is not in the document]
Comment (noah-aviel-dove) [reply scripts/check_branch.py]: I think I found a cleaner way to organize it. Let me know what you think.

**Review by danielsotirhos — APPROVED** [review]
LGTM

File: scripts/check_branch.py [hunk 2]
```diff
@@ -84,15 +89,20 @@ def main(argv):
...
         if stage is None:
             sys.exit(1)
         else:
             print(stage)
     else:
         stage = config.deployment_stage
+        branch = gitlab_branch()
+        if branch is None:
+            if stage == 'sandbox':
+                raise RuntimeError(f"Only the GitLab runner should deploy to '{stage}'")
```
Comment (hannes-ucsc) [inline scripts/check_branch.py:104]: `!r` or `,`

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: scripts/check_branch.py [hunk 3]
```diff
@@ -84,15 +89,20 @@ def main(argv):
...
         if stage is None:
             sys.exit(1)
         else:
             print(stage)
     else:
         stage = config.deployment_stage
+        branch = gitlab_branch()
+        if branch is None:
+            if stage == 'sandbox':
+                raise RuntimeError(f"Only the GitLab runner should deploy to {stage!r}")
```
Comment (hannes-ucsc) [inline scripts/check_branch.py:104]:
```suggestion
                raise RuntimeError(f'Only the GitLab runner should deploy to {stage!r}')
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
For demo, attempt to deploy feature branch to sandbox, and develop to sandbox. Both should fail. Explain how to override the protection.

**Pull request merged**